Alerting: Add optional metadata via query param to silence GET requests

JacobsonMT opened the pull request:
**What is this feature?**

Adds a query param options `ruleMetadata=true` and `accesscontrol=true` to `/alertmanager/grafana/api/v2/silences` and `alertmanager/grafana/api/v2/silence/{SilenceId}`

`ruleMetadata=true` will add a new metadata field containing `rule_uid`, `rule_title`, `folder_uid` for rule-specific silences.
`accessconntrol=true` will add a top-level `accessControl` field containing user permissions on each silence (`read`/`write`/`create`)

Ex:

```json
{
  "accessControl": {
    "create": false,
    "read": true,
    "write": false
  },
  "comment": "test comment",
  "createdBy": "someone",
  "endsAt": "2026-04-18T18:07:26.993Z",
  "id": "e95bcdc3-97a2-44c1-9361-5ec1d12f9f9e",
  "matchers": [{"isEqual": true,"isRegex": false,"name": "__alert_rule_uid__","value": "load_folder0_group3_alert1"}],
  "metadata": {
    "rule_uid": "load_folder0_group3_alert1",
    "rule_title": "group3_alert1",
    "folder_uid": "edmpbowggnm68d"
  },
  "startsAt": "2024-05-24T20:06:18.670Z",
  "status": {
    "state": "active"
  },
  "updatedAt": "2024-05-24T20:06:18.670Z"
}
```

**Why do we need this feature?**

This is needed to:
- Allow the UI to display useful rule information for rule-specific silences without heavy cross-reference API calls.
- Allow the UI to simplify logic on access control for deciding when to display permission-specific components.

JacobsonMT committed:
Alerting: Add optional metadata to GET silence responses

JacobsonMT committed:
Fix typo

JacobsonMT committed:
Simple tests piggybacking on existing silences_test.go

JacobsonMT committed:
Introduce SilencePermissionSet to simplify logic

JacobsonMT commented on pkg/services/ngalert/api/tooling/definitions/alertmanager.go:
```diff
@@ -479,6 +482,50 @@ type GettableSilences = amv2.GettableSilences
+	RuleUID     string                         `json:"rule_uid,omitempty"`
+	RuleTitle   string                         `json:"rule_title,omitempty"`
+	FolderUID   string                         `json:"folder_uid,omitempty"`
+	Permissions map[SilencePermission]struct{} `json:"permissions,omitempty"`
```
Should discuss what format we want to return here. Another option could be `map[SilencePermission]bool` or something else if it's more inline with how we return permissions elsewhere.

tomratcliffe replied:
Yes, I think it'd be more consistent if this returned a map of actions like you said 🙏  This is more consistent with the folder endpoint.

If we wanted to be even more consistent with elsewhere, then the permissions would only be returned if a query param `accesscontrol` was sent - this is how the folder endpoint works, but it doesn't matter too much to me as the UI will be sending the param(s) to get all the metadata on all requests it makes to the silences endpoint

JacobsonMT replied:
Modified the query params and response as you suggested:

Replaces `withMetadata` query param with separate:
- `ruleMetadata`: to request rule metadata.
- `accesscontrol`: to request access control metadata.

Also modifies access control metadata response to be:
`"accessControl": map[string]bool`

Updated the example response in the PR description.

tomratcliffe replied:
Sorry, my fault 🤦 I don't think I was clear enough about the structure - could the metadata/access control object be separate from each other?

So, asking for metadata gives me
```ts
{
  //...normal silence stuff
  id: 'foo',

  metadata: {
    rule_name: 'bar',
  },
}
```

Asking for ACL gives me
```ts
{
  //...normal silence stuff
  id: 'foo',

  accessControl: {
    write: true
    // etc
  }
}
```


So they work independently from each other?

JacobsonMT commented on pkg/services/ngalert/store/alert_rule.go:
```diff
@@ -120,6 +120,29 @@ func (st DBstore) GetAlertRulesGroupByRuleUID(ctx context.Context, query *ngmode
 }
 
+// GetAlertRulesGroupsByRuleUIDs is a handler for retrieving groups of alert rules from that database by UIDs and organisation ID of any rule that belong to the groups.
+func (st DBstore) GetAlertRulesGroupsByRuleUIDs(ctx context.Context, query *ngmodels.GetAlertRulesGroupsByRuleUIDsQuery) (map[ngmodels.AlertRuleGroupKey]ngmodels.RulesGroup, error) {
```
NOTE: This shouldn't be necessary after partial group access changes.

JacobsonMT commented on pkg/services/ngalert/notifier/silence_svc.go:
```diff
@@ -34,43 +48,72 @@ type SilenceStore interface {
 // GetSilence retrieves a silence by its ID.
-func (s *SilenceService) GetSilence(ctx context.Context, user identity.Requester, silenceID string) (*models.Silence, error) {
-	gettableSilence, err := s.store.GetSilence(ctx, user.GetOrgID(), silenceID)
+func (s *SilenceService) GetSilence(ctx context.Context, user identity.Requester, silenceID string, withMetadata bool) (*models.SilenceWithMetadata, error) {
```
NOTE: Another option here is to keep returning `models.Silence` on `GetSilence` and `ListSilences` and instead optionally call `SilenceService.WithMetadata` from the api layer.

JacobsonMT committed:
Modify rule auth checking for new partial group access

JacobsonMT committed:
Separate requests for access control and rule metadata, call from API

Replaces withMetadata query param with separate:
- ruleMetadata: to request rule metadata.
- accesscontrol: to request access control metadata.

Also modifies access control metadata response to be:
`"accessControl": map[string]bool`

The above brings it inline with other access control responses.

JacobsonMT committed:
Remove added code that is no longer used

JacobsonMT committed:
Move accessControl to top-level field in GettableGrafanaSilence

JacobsonMT committed:
Clean ListSilences

JacobsonMT committed:
Tests for accesscontrol and model

yuri-tceretian commented on pkg/services/ngalert/api/api_alertmanager_silences.go:
```diff
@@ -29,7 +31,22 @@ func (srv AlertmanagerSrv) RouteGetSilence(c *contextmodel.ReqContext, silenceID
+	silenceWithMetadata := &models.SilenceWithMetadata{
+		Silence: silence,
+	}
+	if c.QueryBool("accesscontrol") {
```
Also, I wonder if it could be enough to just check existence of parameter, for example `/silences?accesscontrol&metdata`?

JacobsonMT replied:
for `accesscontrol` I'm following the same style as the existing query params for other resources (ex: [Folders](https://github.com/grafana/grafana/blob/1208888bb618774228aefe65884e06316a06e714/pkg/api/folder.go#L432-L434))

yuri-tceretian commented on pkg/services/ngalert/notifier/silence_svc.go:
```diff
@@ -125,3 +143,83 @@ func (s *SilenceService) DeleteSilence(ctx context.Context, user identity.Reques
+			continue // Assume no access if there is an error.
+		}
+
+		if ruleSilences, ok := byRuleUID[rule.UID]; ok {
```
In the case when silence contains ruleUID that does not exist, we do not return any metadata. However, I think it would be better to return "partial" metadata and populate only rule UID. 
This way, UI will be able to detect it as kinda orphaned silence and notify user. WDYT?

yuri-tceretian commented on pkg/services/ngalert/api/api_alertmanager_silences.go:
```diff
@@ -69,3 +98,13 @@ func (srv AlertmanagerSrv) RouteDeleteSilence(c *contextmodel.ReqContext, silenc
 	return response.JSON(http.StatusOK, util.DynMap{"message": "silence deleted"})
 }
+
+func withNoMetadata(silences ...*models.Silence) []*models.SilenceWithMetadata {
```
The name seems to be misleading. Perhaps `withEmptyMetadata`?

yuri-tceretian reviewed (commented):
I tested it and it works as expected. Just few comments\suggestions in comments and
1.  I am not sure `create` field makes sense in per-silence access control field. WDYT?
2. It would be great to update our integration tests. Separate PR is fine, though.

JacobsonMT committed:
Linting and PR comments

JacobsonMT committed:
Fix WithRuleMetadata cache logic

JacobsonMT committed:
Add HasAccessInFolder to ac RuleService fake

JacobsonMT committed:
Add ac FakeSilenceService

JacobsonMT committed:
silence_svc tests

JacobsonMT committed:
Always attach rule uid in rule-specific silence metadata if requested

yuri-tceretian commented on pkg/services/ngalert/notifier/silence_svc.go:
```diff
@@ -125,3 +143,86 @@ func (s *SilenceService) DeleteSilence(ctx context.Context, user identity.Reques
+	}
+
+	if len(permissions) != len(silences) {
+		s.log.Error("failed to get metadata for all silences")
```
perhaps Warn would be enough. WDYT? Error are suitable in a case when the normal program flow was interrupted.

yuri-tceretian reviewed (approved):
LGTM! Really like the direction we're taking!

JacobsonMT committed:
Linting

JacobsonMT committed:
Log Error -> Warn

JacobsonMT committed:
Fix omitempty on GettableGrafanaSilence metadata

tomratcliffe commented:
I checked with my changes from the [UI portion](https://github.com/grafana/grafana/pull/87322) and it looks to be working great on my setup 💯 query params/property names are all aligned now, thanks!
